src/lib/configs/project/dockerCompose.ts

```diff
@@ -0,0 +1,315 @@
+
+  assert(
+    bootStrapNoxNameAndKey !== undefined,
+    "At least one nox is required to generate a docker-compose.yml",
+  );
+
+  const { name: bootStrapNoxName } = bootStrapNoxNameAndKey;
+
+  return {
+    version: "3",
+    services: {
+      [CHAIN_CONTAINER_NAME]: {
+        image: versions.chain,
+        pull_policy: "always",
+        ports: [`${CHAIN_PORT}:${CHAIN_PORT}`],
+      },
+      [IPFS_CONTAINER_NAME]: {
+        image: "ipfs/go-ipfs",
+        pull_policy: "always",
+        ports: [`${IPFS_PORT}:${IPFS_PORT}`, "4001:4001"],
+        environment: {
+          IPFS_PROFILE: "server",
+        },
+        volumes: [`./${IPFS_CONTAINER_NAME}/:/container-init.d/`],
+      },
+      ...Object.fromEntries([
+        genNox({
+          name: bootStrapNoxName,
+          tcpPort: TCP_PORT_START,
+          webSocketPort: WEB_SOCKET_PORT_START,
+          bootStrapNoxName,
+        }),
+      ]),
+      ...Object.fromEntries(
+        restSecretKeys.map(({ name }, index) => {
+          return genNox({
+            name,
+            tcpPort: TCP_PORT_START + 1 + index,
+            webSocketPort: WEB_SOCKET_PORT_START + 1 + index,
+            bootStrapNoxName,
+          });
+        }),
+      ),
+    },
+    secrets: Object.fromEntries(
+      secretKeys.map(({ name, path: file }) => {
+        return [name, { file }];
+      }),
+    ),
+  };
+}
+
+async function genDefaultDockerCompose(
+  providerConfig: ProviderConfigReadonly,
+): Promise<GetDefaultConfig> {
+  const def = await genDockerCompose(providerConfig);
+
+  return () => {
+    return yamlDiffPatch("", {}, def);
+  };
+}
+
+const migrations: Migrations<Config> = [];
+
+type Config = ConfigV0;
+type LatestConfig = ConfigV0;
+export type DockerComposeConfig = InitializedConfig<LatestConfig>;
+export type DockerComposeConfigReadonly =
+  InitializedReadonlyConfig<LatestConfig>;
+
+const initConfigOptions = {
+  allSchemas: [configSchemaV0],
+  latestSchema: configSchemaV0,
+  migrations,
+  name: DOCKER_COMPOSE_FILE_NAME,
+  getConfigOrConfigDirPath: getFluenceDir,
+};
+
+export async function initNewDockerComposeConfig(
+  args: ProviderConfigArgs = {},
+) {
+  const providerConfig = await initNewReadonlyProviderConfig(args);
+  return getConfigInitFunction(
+    initConfigOptions,
+    await genDefaultDockerCompose(providerConfig),
+  )();
+}
+
+export async function initNewReadonlyDockerComposeConfig(
+  args: ProviderConfigArgs = {},
+) {
+  const providerConfig = await initNewReadonlyProviderConfig(args);
+  return getReadonlyConfigInitFunction(
+    initConfigOptions,
+    await genDefaultDockerCompose(providerConfig),
+  )();
+}
+
+export const initDockerComposeConfig = getConfigInitFunction(initConfigOptions);
+
+export const initReadonlyDockerComposeConfig =
+  getReadonlyConfigInitFunction(initConfigOptions);
+
+export const dockerComposeSchema: JSONSchemaType<LatestConfig> = configSchemaV0;
```

Comment:
docs here would be nice. rn the purpose of these consts is unclear to me.

Reply:
I have this in all configs. It's set up like that so the types work when config is extended. `JSONSchemaType<LatestConfig>` is not exactly the same as a `configSchemaV0 as const` type.